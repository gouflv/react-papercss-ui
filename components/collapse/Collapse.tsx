import { useUpdateEffect } from 'ahooks'
import classNames from 'classnames'
import React, {
  FC,
  PropsWithChildren,
  ReactElement,
  ReactNode,
  useEffect,
  useState
} from 'react'
import { isArrayEqual } from '../../utils/array'
import { toArray } from '../../utils/children'

export type CollapseProps = PropsWithChildren<{
  activeKey?: string | string[] | number | number[]
  accordion?: boolean
  onChange?: (activeKey) => void
}>

const Collapse = (props: PropsWithChildren<CollapseProps>) => {
  const [active, setActive] = useState<string[]>([])

  useEffect(() => {
    if (typeof props.activeKey === 'undefined') return
    if (Array.isArray(props.activeKey)) {
      setActive((props.activeKey as any[]).map(k => String(k)))
    } else {
      setActive([String(props.activeKey)])
    }
  }, [props.activeKey])

  function childRender() {
    return toArray<ReactElement>(props.children).map((child, index) => {
      const key = child.props.name || String(index)
      const isActive = !!~active.indexOf(key)
      return React.cloneElement(child, {
        ...child.props,
        key,
        name: child.props.name || key,
        active: isActive,
        onClick: onPanelClick
      } as PanelProps)
    })
  }

  function onPanelClick(key: string | number) {
    const _active = String(key)
    if (props.accordion) {
      setActive(!!~active.indexOf(_active) ? [] : [String(key)])
    } else {
      setActive(prevState => {
        const keySet = new Set(prevState)
        keySet.has(_active) ? keySet.delete(_active) : keySet.add(_active)
        return Array.from(keySet)
      })
    }
  }

  useUpdateEffect(() => {
    if (typeof props.onChange !== 'function') return
    if (props.accordion) {
      if (active !== props.activeKey) {
        props.onChange(active)
      }
    } else {
      if (
        !isArrayEqual(
          active,
          Array.isArray(props.activeKey) ? props.activeKey : [props.activeKey]
        )
      ) {
        props.onChange(active)
      }
    }
  }, [active])

  return <div className='collapse-group'>{childRender()}</div>
}

export interface PanelProps {
  name?: number | string
  header: ReactNode
  active?: boolean
  onClick?: (key: string | number) => void
}

const Panel: FC<PanelProps> = props => {
  function onClick() {
    if (props.onClick && props.name) {
      props.onClick(props.name)
    }
  }

  return (
    <div
      className={classNames('collapsible', {
        'collapsible--active': props.active
      })}
      data-name={props.name}
    >
      <label onClick={onClick}>{props.header}</label>
      <div className='collapsible-body'>{props.children}</div>
    </div>
  )
}

Collapse.Panel = Panel

export { Collapse }
