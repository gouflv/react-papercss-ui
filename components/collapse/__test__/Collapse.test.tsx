import { render, screen } from '@testing-library/react'
import React, { PropsWithChildren } from 'react'
import { Collapse, CollapseProps } from '../Collapse'

const { Panel } = Collapse

describe('Test Collapse Component', () => {
  let props: PropsWithChildren<CollapseProps>

  beforeEach(() => {
    props = {
      children: [
        <Panel key='1' name='1' header='Fist'>
          Fist Panel
        </Panel>,
        <Panel key='2' name='2' header='Second'>
          Second Panel
        </Panel>
      ]
    }
  })

  const renderComponent = () => render(<Collapse {...props} />)

  it('should open panel with active set as "1"', () => {
    props = {
      ...props,
      activeKey: '1'
    }
    const { container } = renderComponent()
    expect(container).toMatchSnapshot()
    expect(screen.getByText('Fist Panel')).toBeDefined()
  })

  it('should open un-name panel with active set as "1"', () => {
    props = {
      activeKey: 0,
      children: [
        <Panel key='1' header='Fist'>
          Fist Panel
        </Panel>,
        <Panel key='2' header='Second'>
          Second Panel
        </Panel>
      ]
    }
    const { container } = renderComponent()
    expect(container).toMatchSnapshot()
  })
})
