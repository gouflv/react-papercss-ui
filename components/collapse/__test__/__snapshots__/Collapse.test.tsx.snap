// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Collapse Component should open panel with active set as "1" 1`] = `
<div>
  <div
    class="collapse-group"
  >
    <div
      class="collapsible"
    >
      <label>
        Fist
      </label>
      <div
        class="collapsible-body"
        style="max-height: none; padding: 0.75rem; opacity: 1;"
      >
        Fist Panel
      </div>
    </div>
    <div
      class="collapsible"
    >
      <label>
        Second
      </label>
    </div>
  </div>
</div>
`;

exports[`Test Collapse Component should open un-name panel with active set as "1" 1`] = `
<div>
  <div
    class="collapse-group"
  >
    <div
      class="collapsible"
    >
      <label>
        Fist
      </label>
    </div>
    <div
      class="collapsible"
    >
      <label>
        Second
      </label>
    </div>
  </div>
</div>
`;
